### PR TITLE
Refactor GLSL shader compilation

### PR DESCRIPTION
**Pull request by reduz**
* Used a more consistent set of keywords for the shader
* Remove all harcoded entry points
* Re-wrote the GLSL shader parser, new system is more flexible. Allows any entry point organization (need this for particles)
* Entry point for sky shaders is now sky().
* Entry point for particle shaders is now process().

image

Warning: Visual shaders (some) may have broken, will need fixing. Probably better to wait until I am done fixing particles in subsequent PRs.

### CODE REVIEW

File: servers/rendering/renderer_rd/shader_compiler_rd.cpp [hunk 1]
```diff
@@ -858,7 +842,7 @@ String ShaderCompilerRD::_dump_node_code(const SL::Node *p_node, int p_level, Ge
 			SL::VariableNode *vnode = (SL::VariableNode *)p_node;
 			bool use_fragment_varying = false;
 
-			if (current_func_name != vertex_name) {
+			if (!(p_actions.entry_point_stages.has(current_func_name) && p_actions.entry_point_stages[current_func_name] != STAGE_VERTEX)) {
```
Comment (reduz) [inline servers/rendering/renderer_rd/shader_compiler_rd.cpp:845]: hmm I think this is wrong..

File: servers/rendering/renderer_rd/shaders/canvas.glsl [hunk 2]
```diff
@@ -26,16 +26,16 @@ layout(location = 3) out vec2 pixel_size_interp;
...
 #endif
 
-#ifdef USE_MATERIAL_UNIFORMS
+#ifdef MATERIAL_UNIFORMS_USED
 layout(set = 1, binding = 0, std140) uniform MaterialUniforms{
-	/* clang-format off */
-MATERIAL_UNIFORMS
+/* clang-format off */
+#MATERIAL_UNIFORMS
 	/* clang-format on */
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/shaders/canvas.glsl:33]: Formatting of comment

File: servers/rendering/renderer_rd/shaders/sky.glsl [hunk 3]
```diff
@@ -202,20 +202,10 @@ void main() {
...
-// unused, just here to make our compiler happy, make sure we don't execute any light code the user adds in..
-#ifndef REALLYINCLUDETHIS
 	{
-		/* clang-format off */
-
-LIGHT_SHADER_CODE
-
-		/* clang-format on */
-	}
-#endif
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/shaders/sky.glsl:214]: Nice to see that we can delete this now.

File: servers/rendering/shader_language.h [hunk 4]
```diff
@@ -769,7 +780,8 @@ class ShaderLanguage {
 		Map<StringName, BuiltInInfo> built_ins;
 		Map<StringName, StageFunctionInfo> stage_functions;
 
-		bool can_discard;
+		bool can_discard = false;
+		bool main_function = false;
```
Comment (clayjohn) [inline servers/rendering/shader_language.h:784]: The concept of "main function" is a little confusing for me. Why are "fragment", "vertex", and "light" "main functions", but "process" and "sky" are not?

This flag seems to flag functions that can pass data to other functions. So maybe a different name would make that clear?

File: servers/rendering/renderer_rd/shaders/canvas.glsl (hunk 2 above)
Comment (reduz) [reply servers/rendering/renderer_rd/shaders/canvas.glsl:33]: I wonder if these clang-format rules are needed any longer now that I added a #.. @akien-mga?

File: servers/rendering/renderer_rd/shaders/sky.glsl (hunk 3 above)
Comment (reduz) [reply servers/rendering/renderer_rd/shaders/sky.glsl:214]: oh.. yeah :sweat:

File: servers/rendering/shader_language.h (hunk 4 above)
Comment (reduz) [reply servers/rendering/shader_language.h:784]: They should be, unless I messed up, need to give it a check.

Comment (reduz) [reply servers/rendering/shader_language.h:784]: Main function is just a hint to the compiler, mostly that it should not return anything. SubFunctions don't get this set.

File: servers/rendering/renderer_rd/shader_rd.h [hunk 5]
```diff
@@ -76,40 +75,53 @@ class ShaderRD {
...
+	struct StageTemplate {
+		struct Chunk {
+			enum Type {
+				TYPE_VERSION_DEFINES,
+				TYPE_MATERIAL_UNIFORMS,
+				TYPE_VERTEX_GLOBALS,
+				TYPE_FRAGMENT_GLOBALS,
+				TYPE_COMPUTE_GLOBALS,
+				TYPE_CODE,
+				TYPE_TEXT
```
Comment (aaronfranke) [inline servers/rendering/renderer_rd/shader_rd.h:87]:
```suggestion
				TYPE_TEXT,
```

**Review by JFonS — COMMENTED** [review]
There are some unnecessary (I think) scopes around the various `shader_singleton->shader.version_set_code()` calls, just a cosmetic nitpick.

Other than that, as @clayjohn mentions `process()` and `sky()` don't seem to have `main_function` set to true.

The rest looks good :)

**Comment by reduz** [comment]
@JFonS ah right, I started by passing a structure there that then I removed so I guess those remained as-is, no reason to have them now.

**Review by JFonS — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**